### PR TITLE
Add check for both init= = or none

### PR DESCRIPTION
**Pull request by mppf**
Related to issue #8065.

Previously, the compiler generated `init=` and `=` for all types, but did
so separately. That means that one could use the compiler-generated `=`
but a custom `init=` or vice versa. This could in some cases lead to
confusing and hard to debug problems if a custom `init=` or `=` was
forgotten or included the wrong type signature.

To resolve these problems, this PR started by adjusting the compiler to
check that if a type has custom `init=`, it does not use the default `=`;
and that if a type has a custom `=`, it does not use the default `init=`.

The compiler currently only considers the result type (`this` for `init=`
and the LHS for `=`) when doing this check. As a result, a type with an
`init=` or `=` function with a different type RHS causes the error unless
both `=` and `init=` are added for the same type. This does not seem to
be an undue burden but could have an impact on determining which types
are POD. I recommend that in the future types can opt in to being
considered POD (in which case `=` and `init=` will be resolved, and
compilerErrors in them issued, but can be removed during optimization).
The generalized attribute syntax #14141 would be a reasonable way to do
that. Even before this PR, the `range` type needed to opt in to being POD
because its `init=` included various checks.

Next, this PR adjusted the compiler to also check for a mix of uses of
compiler-generated `init=` and custom `init=`; or compiler-generated `=`
and custom `=`. This only required adjusting 2 additional tests beyond
the tests adjusted for the above rule. In all, this PR required adjusting
~30 tests / modules and almost all of these were tests of record
initializers/assignment.

Reviewed by @lydia-duncan - thanks! Thanks to many others for discussion
about the language design direction.

- [x] full local futures testing

### CODE REVIEW

File: compiler/main/checks.cpp [hunk 1]
```diff
@@ -700,6 +703,86 @@ checkAutoCopyMap()
   }
 }
 
+static FnSymbol* findUserInitEq(AggregateType* at) {
+  for_alive_in_Vec(FnSymbol, fn, gFnSymbols) {
+    if (fn->name == astrInitEquals &&
+        !fn->hasFlag(FLAG_COMPILER_GENERATED) &&
+        fn->numFormals() >= 2) {
```
Comment (lydia-duncan) [inline compiler/main/checks.cpp]: Do we allow `init=` with more than two arguments, or do we catch them elsewhere?

File: compiler/main/checks.cpp [hunk 2]
```diff
@@ -700,6 +703,86 @@ checkAutoCopyMap()
...
+
+
+static void
+checkDefaultInitEqAndAssign()
+{
+  for_alive_in_Vec(TypeSymbol, ts, gTypeSymbols) {
+    if (ts->hasFlag(FLAG_EXTERN))
+      continue; // can't make init= for extern types today anyway
+    if (!isRecord(ts->type) && !isUnion(ts->type))
+      continue; // can't make init= for non-records today anyway
```
Comment (lydia-duncan) [inline compiler/main/checks.cpp]: That comment is incorrect, since it looks like it allows unions

File: compiler/main/checks.cpp [hunk 3]
```diff
@@ -700,6 +703,86 @@ checkAutoCopyMap()
...
+    AggregateType* at = toAggregateType(ts->type);
+    bool defaultInitEq = ts->hasFlag(FLAG_TYPE_DEFAULT_INIT_EQUAL);
+    bool customInitEq = ts->hasFlag(FLAG_TYPE_CUSTOM_INIT_EQUAL);
+    bool defaultAssign = ts->hasFlag(FLAG_TYPE_DEFAULT_ASSIGN);
+    bool customAssign = ts->hasFlag(FLAG_TYPE_CUSTOM_ASSIGN);
+    if (defaultInitEq && customAssign) {
+      USR_FATAL_CONT(ts, "Type '%s' uses compiler-generated default init= "
+                         "but has a custom = function. "
+                         "Please add an init= method",
+                         toString(ts->type));
```
Comment (lydia-duncan) [inline compiler/main/checks.cpp]: Should these error messages have a ` around the init= and = portions of it?

File: test/types/records/ferguson/must-init-record/myrecord.chpl [hunk 4]
```diff
@@ -53,47 +53,19 @@ proc R.verify() {
...
-
-  // allow copies of a default-initialized record.
-  ret.setup(x = arg.x, true);
-
-  if debug {
-    writeln("leaving init copy");
-  }
-
-  return ret;
 }
```
Comment (lydia-duncan) [inline test/types/records/ferguson/must-init-record/myrecord.chpl]: I love that this gets rid of an initCopy/autoCopy

File: compiler/main/checks.cpp (hunk 1 above)
Comment (mppf) [reply compiler/main/checks.cpp]: I meant to just check `numFormals >= 1` - I just didn't want to cause core dumps for erroneous `init=` here.

For `init=` with more arguments, I'd expect something like a default argument would be allowed. But that's not really the main issue in this PR. This routine in particular only exists for reporting error messages.

File: test/types/records/ferguson/must-init-record/myrecord.chpl (hunk 4 above)
Comment (mppf) [reply test/types/records/ferguson/must-init-record/myrecord.chpl]: I think of this more as a related test change - we could have done this before...

File: compiler/main/checks.cpp (hunk 3 above)
Comment (mppf) [reply compiler/main/checks.cpp]: I put single quotes like other error messages. It's not obvious to me that's better but I interpreted your comment as a suggestion.

**Pull request merged**